rclcpp/include/rclcpp/publisher_options.hpp

```diff
@@ -113,9 +113,26 @@ struct PublisherOptionsWithAllocator : public PublisherOptionsBase
   }
 
 private:
+  using PlainAllocator =
+    typename std::allocator_traits<Allocator>::template rebind_alloc<char>;
+
+  rcl_allocator_t
+  get_rcl_allocator() const
+  {
+    if (!plain_allocator_storage_) {
+      plain_allocator_storage_ =
+        std::make_shared<PlainAllocator>(*this->get_allocator());
+    }
+    return rclcpp::allocator::get_rcl_allocator<char>(*plain_allocator_storage_);
+  }
+
   // This is a temporal workaround, to make sure that get_allocator()
   // always returns a copy of the same allocator.
   mutable std::shared_ptr<Allocator> allocator_storage_;
+
+  // This is a temporal workaround, to keep the plain allocator that backs
+  // up the rcl allocator returned in rcl_publisher_options_t alive.
+  mutable std::shared_ptr<PlainAllocator> plain_allocator_storage_;
```

Comment:
can this be

```suggestion
  mutable std::optional<PlainAllocator> plain_allocator_storage_;
```

?

Reply:
Yeah, `std::shared_ptr` makes no sense in this context. Perhaps `std::unique_ptr`, so we can backport to Foxy?

Reply:
Well, neither `std::shared_ptr` nor `std::optional` work in a situation like:

> it'd be easy to fetch rcl_publisher_options_t from some PublisherOptions, then copy those options, then drop the original

Arguably, that's bogus code (lifetimes are not properly managed). Alas, https://github.com/ros2/rclcpp/blob/1fc2d58799a6d4530522be5c236c70be53455e60/rclcpp/include/rclcpp/subscription.hpp#L115-L118. `std::shared_ptr` would still work in that case, but only because the execution order is favorable.